source/k64f_eth_nanostack_port.c

```diff
@@ -501,236 +361,275 @@ static int8_t arm_eth_phy_k64f_address_write(phy_address_type_e address_type, ui
+ *      Eth_IRQ_Thread (thread context) after receiving the signal from
+ *      interrupt context (SIG_TX).
+ *      Otherwise, it will be called directly from the interrupt context.
+ */
+static void enet_tx_task(void)
+{
+    tx_queue_reclaim();
 }
 
-/* Transmit IRQ Handler*/
-void ENET_Transmit_IRQHandler(void)
+/** \brief  Callback function to receive Ethernet TX/RX Events
+ *
+ *  Called from the interrupt context. If we are not using mbed RTOS, we call up
+ *  the ISRs, otherwise we send a corresponding signal to the responsible Thread.
+ *
+ *  \param[in] base  Pointer to base structure of Eth module
+ *  \param[in] handle Pointer to global ethernet handle
+ *  \param[in] event  Type of interrupt or any other event (what we have
+ *             registered in init)
+ *  \param[in] param A pointer to user provided data|parameter. [optional]
+ */
+static void ethernet_event_callback(ENET_Type *base, enet_handle_t *handle, enet_event_t event, void *param)
 {
-    NVIC_SetPendingIRQ(ENET_Receive_IRQn);
+    switch (event) {
+        case kENET_RxEvent:
+#ifdef MBED_CONF_RTOS_PRESENT
+            osSignalSet(eth_irq_thread_id, SIG_RX);
+#else
+            enet_rx_task();
+#endif /*MBED_CONF_RTOS_PRESENT*/
+            break;
+        case kENET_TxEvent:
+#ifdef MBED_CONF_RTOS_PRESENT
+            osSignalSet(eth_irq_thread_id, SIG_TX);
+#else
+            enet_tx_task();
+#endif /*MBED_CONF_RTOS_PRESENT*/
+            break;
+        default:
+            break;
+    }
 }
 
-/* Receive IRQ Handler*/
-void ENET_Receive_IRQHandler(void)
+#ifdef MBED_CONF_RTOS_PRESENT
+/** \brief  Task started by ETH_IRQ_Thread
+ *
+ *      Used only in case of  mbed RTOS. This task handles the signals coming
+ *      from interrupt context.
+ *
+ *  \param[in] Optional user-given parameter
+ */
+static void Eth_IRQ_Task(const void *x)
 {
-    enet_dev_if_t *ethernet_iface_ptr =
-            &ethernet_iface[BOARD_DEBUG_ENET_INSTANCE];
+    for (;;) {
+        osEvent event =  osSignalWait(0, osWaitForever);
+        if (event.status != osEventSignal) {
+            continue;
+        }
 
-    platform_interrupts_disabled();
-    if (enet_hal_get_interrupt_status(ethernet_iface_ptr->deviceNumber,
-                                      kEnetRxFrameInterrupt)) {
-        enet_mac_rx_isr(enetIfHandle);
-    }
+        eth_if_lock();
 
-    if (enet_hal_get_interrupt_status(ethernet_iface_ptr->deviceNumber,
-                                      kEnetTxFrameInterrupt)) {
-        enet_mac_tx_isr(enetIfHandle);
+        if (event.value.signals & SIG_RX) {
+            enet_rx_task();
+        }
+
+        if (event.value.signals & SIG_TX) {
+            enet_tx_task();
+        }
+
+        eth_if_unlock();
     }
-    platform_interrupts_enabling();
 }
 
 
+/** \brief  Function creating the IRQ thread
+ *
+ *      Used only in case of  mbed RTOS. Creates a thread for IRQ task.
+ */
+static void Eth_IRQ_Thread(void)
```

Comment:
Naming is confusing - this isn't the thread. It starts the thread. Eth_IRQ_Task is the thread.

In various comments you also talk about routines being called from "Eth_IRQ_Thread", which is wrong.